### ISSUE TITLE
some bugs in the mqMorphoDigCore.cxx

### ISSUE THREAD

**Issue by Garysk** [morphomuseum/MorphoDig#12]
First, I am poor about english writing.
    MorphoDig is a good project. But when I downloaded the master branch and build it using debug mode, it crashed after the Volume rendering mask operation.
    I found some bugs in MorphoDig-master\MorphoDig\Qt\mqMorphoDigCore.cxx. They are all about double pointWC_xxx[3] variables  and the GetDisplayToWorld() function.
    The details are below:
     1) line 2064: 
         double pointWC_center[3] = {0, 0, 0};
         line 2067:
         this->GetDisplayToWorld(ptCenter[0], ptCenter[1], ptCenter[2], pointWC_center[3]);
      2) line 2215/2216/13831/13832/14272/14273
          double pointWC_near[3] = { 0, 0, 0 };
	  double pointWC_far[3] = { 0,0,0 }; 
      3) The function GetDisplayToWorld's 4th parameter should be double worldPt[4]，and there is worldPt[3]=1.0 line in the function setting the 4th element。The augments pointWC_center/pointWC_near/pointWC_far passed to GetDisplayToWorld have only 3 elements. It caused array bounds errors and made the program compiled in debug mode crashed。

**Comment by renaud-lebrun** [comment]
Wow thanks a lot, this is probably why from my side it crashed from time to time when using masking options!!!

**Comment by renaud-lebrun** [comment]
I just committed a new mqMorphoDigCore.cxx that should fix the problem.
Thanks again!
=> can you tell me if it fixes your compilation issues ? 
=> just curious : on which OS do you try to compile and use MorphoDig? I seldom have feedback on that topic...

**Comment by Garysk** [comment]
thanks a lots for your answer.  My environments include win 10 + qt 5.15.2 + vtk 9. 0. 1 + MorphoDig master + vs2019 community . 
1) qt build from source following official how to 
2)  vtk build from source following MorphoDig's how to 
3)  MorphoDig build from source 
4)  all build using debug mode
5)  encounter three questions 
      i) MorphoDig's source code page is 1252 or 1250, my code page is 936.  This really made big problem.  I took about one day to find it and converted all cxx/h file to utf8 format using iconv from gnuwin32. 
    Ii) delete vtkSetGet.h(vtk's header) line 696.  This code line is only used for debug and caused a lot of compile errors. The code vtkSetMacro(Box, vtkSmartPointer<vtkBoxWidget>) in vtkMDActor. h and some other MorphoDig code files . 
   iii) The bugs I had reported. 
Now the program runs normally.  Thanks again.

**Comment by Garysk** [comment]
Not to delete vtkSetMacro(Box,  vtkSmartPointer<vtkBoxWidget >).  I mean this code causing compile errors. delete line 696 of vtkSetGet.h can solve this problem. Sorry for my  poor English.

**Comment by renaud-lebrun** [comment]
Thanks!

<i) MorphoDig's source code page is 1252 or 1250, my code page is 936. This really made big problem. I took about one day to find it and converted all cxx/h file to utf8 format using iconv from gnuwin32.
=> MorphoDig's source code page is 1252 : which file ? mqMorphoDigCore.cxx?

Strange, when I open a few cxx/h files with notepad++ (among which mqMorphoDigCore.cxx, I find that these files are encoded already in UTF8. Could you tell me which files were not UTF8, and in which format they were encoded ? 

<Ii) delete vtkSetGet.h(vtk's header) line 696. 
In which file I can't find that ?

Thanks again for your help 
Best wishes,
Renaud.

**Comment by Garysk** [comment]
you are welcome. I see the email just now. Sorry for the late reply。
1) the vtkSetGet.h file's full path in vtk9.0.1 is VTK-9.0.1\Common\Core\vtkSetGet.h. It's line 696 is vtkmsg << "" x; . This made visual studio complaining about multiple overloaded function。The root is code like vtkSetMacro(Hist, vtkSmartPointer<vtkImageAccumulate>); in  MorphoDig-master\MorphoDig\Helpers\vtkMDVolume.h(line 156) and in other lines and files, because vtkmsg's type is vtkOStrStreamWrapper vtkmsg; (line 687 in vtkSetGet.h) , and vtkOStrStreamWrapper has no overloaded operator << for type vtkSmartPointer<T>， so the visual studio complained。I did not want to add some overloaded operator << in class vtkOStrStreamWrapper, and vtkmsg << "" x; is only for debug(release version has not the problem), so I delete it simply.
The reason is too complex for me to describe it clearly in english. In a word, Something like that.

2) One disorderly code file when open with utf8 code is MorphoDig\Qt\mqCameraReaction.cxx(line 38). Maybe some other files. I were not verify all files。The image below shows you the status。The code problem caused the next line jump up to the end of the disorderly code line, and visual studio 2019 communit complained the variable camdist was not defined. 


The code is normally when open with 1252. I catch a image too.


My sincerest appreciations to you !
Best wish!
shengke gan
















At 2021-03-08 16:54:56, "renaud-lebrun" <notifications@github.com> wrote:

Thanks!

<i) MorphoDig's source code page is 1252 or 1250, my code page is 936. This really made big problem. I took about one day to find it and converted all cxx/h file to utf8 format using iconv from gnuwin32.
=> MorphoDig's source code page is 1252 : which file ? mqMorphoDigCore.cxx?

Strange, when I open a few cxx/h files with notepad++ (among which mqMorphoDigCore.cxx, I find that these files are encoded already in UTF8. Could you tell me which files were not UTF8, and in which format they were encoded ?

<Ii) delete vtkSetGet.h(vtk's header) line 696.
In which file I can't find that ?

Thanks again for your help
Best wishes,
Renaud.

**Comment by renaud-lebrun** [comment]
Thanks a lot!
#1: it might be a duplication of macro declaration in MorphoDig code, I have to find out, I think that I remember that at some point I  had redeclared the set and get macro code; I do not remember why I did that though. Anyway native VTK files should not have to be modified... I will check that.
#2 mqCameraReaction.cxx : I do not get it sorry, from my side this file is UTF8, so I do not know what I can do to fix this issue... 

Thanks again for your help.

**Comment by Garysk** [comment]
You are really a responsible author.
I don't known the email would be send to this github comment page. I took two snapshots and send with the answer email, but they disappeared.
The two tissues are all unimportant and can be easy to overcome. 
The first tissue only affects the debug version and not exists in the release version.
The second tissue may be repeated if download the files from github. I downloaded the project zip file and found the disorderly code when i opened it using visual studio code with utf8 .  But I really do not think it's a serious tissue.
You are very kindly and patient. Thank you very much.
